Mathlib/RingTheory/Coalgebra.lean

```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2023 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey
+-/
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Definition and example of a coalgebra
+-/
+
+universe u v
+
+open scoped TensorProduct
+
+/-- A coalgebra over a commutative ring `R` is a module over `R` equipped with a coassociative
+comultiplication `Δ` and a counit `ε` obeying the left and right conunitality laws. -/
+class Coalgebra (R : Type u) (A : Type v) [CommRing R] [AddCommGroup A] [Module R A] where
+  /-- The comultiplication of the coalgebra -/
+  Δ : A →ₗ[R] A ⊗[R] A
+  /-- The counit of the coalgebra -/
+  ε : A →ₗ[R] R
+  /-- The comultiplication is coassociative -/
+  coassoc : TensorProduct.assoc R A A A ∘ₗ TensorProduct.map Δ .id ∘ₗ Δ =
+    TensorProduct.map .id Δ ∘ₗ Δ
+  /-- The counit satisfies the left counitality law -/
+  ε_id : TensorProduct.lid R A ∘ₗ TensorProduct.map ε .id ∘ₗ Δ = .id
+  /-- The counit satisfies the right counitality law -/
+  id_ε : TensorProduct.rid R A ∘ₗ TensorProduct.map .id ε ∘ₗ Δ = .id
+
+namespace Coalgebra
+
+section CommRingAddCommGroup
+
+variable {R : Type u} {A : Type v}
+variable [CommRing R] [AddCommGroup A] [Module R A] [Coalgebra R A]
+
+@[simp]
+theorem coassoc_apply (a : A) : TensorProduct.assoc R A A A (TensorProduct.map Δ .id (Δ a)) =
+    TensorProduct.map .id Δ (Δ a) := LinearMap.congr_fun coassoc a
+
+@[simp]
+theorem ε_id_apply (a : A) : TensorProduct.lid R A (TensorProduct.map ε .id (Δ a)) = a :=
+    LinearMap.congr_fun ε_id a
+
+@[simp]
+theorem id_ε_apply (a : A) : TensorProduct.rid R A (TensorProduct.map .id ε (Δ a)) = a :=
+    LinearMap.congr_fun id_ε a
+
+end CommRingAddCommGroup
+
+end Coalgebra
+
+namespace Finsupp
+
+section CommRing
+
+variable (R : Type u) (S : Type v)
+variable [CommRing R]
+
+/-- The `R`-module whose elements are functions `S → R` which are zero on all but finitely many
+elements of `S` has a coalgebra structure. The coproduct is given by `Δ(fₛ) = fₛ ⊗ fₛ` and the
+counit by `ε(fₛ) =  1`, where `fₛ` is the function sending `s` to `1` and all other elements of `S`
+to zero. -/
+noncomputable
+instance instCoalgebra : Coalgebra R (S →₀ R) where
+  Δ := Finsupp.total S ((S →₀ R) ⊗[R] (S →₀ R)) R
+    (fun s ↦ Finsupp.single s 1 ⊗ₜ Finsupp.single s 1)
+  ε := Finsupp.total S R R (fun _ ↦ 1)
+  coassoc := by
+    ext; simp
+  ε_id := by
+    ext; simp
+  id_ε := by
+    ext; simp
+
+theorem Δ_single_one (s : S) : (instCoalgebra R S).Δ (Finsupp.single s 1) =
+    (Finsupp.single s 1) ⊗ₜ[R] (Finsupp.single s 1) := by
+  unfold Coalgebra.Δ; unfold instCoalgebra; simp
+
+theorem ε_single_r (s : S) (r : R) : (instCoalgebra R S).ε (Finsupp.single s r) = r := by
```

Comment:
Theorem names shouldn't mention the names of their variables, because the caller doesn't care about those:
```suggestion
theorem ε_single (s : S) (r : R) : (instCoalgebra R S).ε (Finsupp.single s r) = r := by
```